benchmarks/deconvolution-1D/server.py

```diff
@@ -0,0 +1,139 @@
+import umbridge
+import numpy as np
+import cuqi
+
+
+class Deconvolution1D(umbridge.Model):
+    """Base benchmark for all 1D deconvolution problems"""
+
+    dim = 128  # Dimension of the 1D signal
+    default_delta = 0.01  # Default value for delta
+
+    def __init__(self, name):
+        """Initialize the model
+
+        parameters
+        ----------
+        name : str
+            Name of the model
+
+        """
+        # Load data and exact phantom from file.
+        # These were originally generated by the testproblem in cuqipy,
+        # with the testproblem parameters shown below.
+        # np.random.seed was set to 0.
+        data = np.load("data_square.npz")
+
+        # Set up test problem
+        TP = cuqi.testproblem.Deconvolution1D(
+            dim=self.dim,
+            kernel="gauss",
+            kernel_param=10,
+            phantom="square",
+            data=data["data"], # data was None (i.e. generated from exact phantom).
+            noise_std=0.05,
+            noise_type="gaussian",
+        )
+
+        self.likelihood = TP.likelihood
+        self.prior = TP.prior
+
+        super().__init__(name)
+
+    def get_input_sizes(self, config):
+        return [self.dim]
+
+    def get_output_sizes(self, config):
+        return [1]
+
+    def __call__(self, parameters, config):
+        posterior = self._configure_posterior(config)
+        output = posterior.logpdf(np.asarray(parameters[0]))
+        return [[output[0]]]
+
+    def gradient(self, out_wrt, in_wrt, parameters, sens, config):
```

Comment:
Our current implementation of `.gradient` in CUQIpy evaluates the gradient of the logpdf function of the posterior. Since the posterior is always R^n -> R, we simply need to evaluate on the same input as the input given to the logpdf function.

It was my understanding that this was `parameters`? Am I wrong in this?

Reply:
Sorry, I think I put that in a confusing way.
You are right in passing `parameters` to CUQIpy's `.gradient`. The gradient in UM-Bridge here should just take CUQIpy's `.gradient` and multiply that by `sens[0]`.

Reply:
I am afraid I do not really understand. The returned value by CUQIpy gradient is a vector representing the gradient of the logpdf w.r.t. the parameters. How and why do you want me to multiply it with sens? Elementwise?

Reply:
Yes, should be elementwise. The idea is that we don't expose the full Jacobian `J` in UM-Bridge, but instead "jacobian action" `J v` or "gradient" `v^T J`. This interface is basically the same as in MUQ, there is some motivation for it documented [here](https://mituq.bitbucket.io/source/_site/latest/classmuq_1_1Modeling_1_1ModPiece.html#a15a6520aba1be966b9d40d843c4ead8d).
This is a bit overkill in case of gradient of a logpdf, but makes sense for larger inputs/outputs or PDE solves that can efficiently give you these operations but not the full Jacobian.

Reply:
Thank you Linus, I now understand. We do the same in CUQIpy. The gradient returns the Jacobian action. For succintness the user can choose not to pass the sens when the jacobian action is independent of it. I will make it more clear by explicitly passing it to CUQIpy. Hope that clears things up.

Reply:
Sounds great, thanks! :)

Reply:
Hi @linusseelinger. I updated the gradient computation now with some comments. I hope I got it right. Could you maybe do a check that the gradient method of the umbridge model returns the expected output by e.g. checking against a numerical result?

Reply:
Made a small modification to take into account sens. The gradient now passes pymc's gradient check (see pymc example client if you'd like to reproduce, you only have to change the input dimension to 128).

Reply:
Ok thanks